tests/project_helpers.py

```diff
@@ -0,0 +1,245 @@
+            if expected_names is None:
+                print(f"Skip checking .{field.name}")
+                continue
+
+            print(f"Checking .{field.name}")
+            actual_names = member_extractor(field.name)
+            print(f"  Expect names: {sorted(expected_names)}")
+            print(f"  Actual names: {sorted(actual_names)}")
+            assert expected_names == actual_names
+
+    def verify_analysis(self, analysis: Analysis) -> None:
+        """Assert that the given Analysis object matches our expectations."""
+        self._verify_fields(lambda member: {m.name for m in getattr(analysis, member)})
+
+    def verify_analysis_json(self, analysis: JsonData) -> None:
+        """Assert that the given JSON analysis matches our expectations."""
+        self._verify_fields(lambda member: {m["name"] for m in analysis[member]})
+
+
+@dataclass
+class BaseExperiment:
+    """A single experiment, running FawltyDeps on a test project.
+
+    An experiment is part of a bigger project (see BaseProject below) and has:
+    - A name and description, for documentation purposes.
+    - A list of requirements, to be installed into a virtualenv and made
+      available to FawltyDeps when this experiment is run
+      (see CachedExperimentVenv for details).
+    - A set of expectations on the resulting Analysis object, to be verified
+      after the FawltyDeps has been run (see AnalysisExpectations for details).
+    """
+
+    name: str
+    description: Optional[str]
+    requirements: List[str]
+    expectations: AnalysisExpectations
+
+    @staticmethod
+    def init_args_from_toml(name: str, data: TomlData) -> Dict[str, Any]:
+        return dict(
+            name=name,
+            description=data.get("description"),
+            requirements=data.get("requirements", []),
+            expectations=AnalysisExpectations.parse_from_toml(data),
+        )
+
+    @classmethod
+    def parse_from_toml(cls, name: str, data: TomlData) -> "BaseExperiment":
+        raise NotImplementedError
+
+    def get_venv_dir(self, cache: pytest.Cache) -> Path:
+        """Get this venv's dir and create it if necessary."""
+        venv = CachedExperimentVenv(self.requirements)
+        return venv(cache)
+
+
+@dataclass
+class BaseProject:
+    """Encapsulate a Python project to be tested with FawltyDeps.
+
+    This represents project on which we want to run FawltyDeps in one or more
+    experiments. It has at least:
+    - A name and description, for documentation purposes.
+    - A list of experiments (see BaseExperiment above), describing one or more
+      scenarios for running FawltyDeps on this project, and what results to
+      expect in those scenarios.
+    """
+
+    name: str
+    description: Optional[str]
+    experiments: List[BaseExperiment]
+
+    @staticmethod
+    def parse_toml(
+        toml_path: Path, ExperimentClass: Type[BaseExperiment]
+    ) -> Tuple[Dict[str, Any], TomlData]:
```

Comment:
I would expect that parsing tokml will result in a `BaseProject` object.

Reply:
This threw me / caused me a bit of pause at first, too, but it made a little more sense once I reconsidered and viewed it as an internal helper. I think a couple things would help lessen surprise here, though.
a) We strive to name members that parse some file or preprocessed data _and_ return an instance of the containing type as `from_*` (rather than `parse_*` or `parse_from_*`) and reserve the `parse_*`-like naming for helpers.
b) This appears to be a method called only from subclasses, so we could prefix with an underscore to reinforce the idea that it's a helper rather than an API member like a `from_*` method tagged as `classmethod` would be expected to be.

Reply:
Yeah, I tried that at first, but I'm not sure how to do a multi-stage construction like that with inherited dataclasses: The base class does not really exist as an instance separate from the subclass; when I construct a `SampleProject` or `ThirdPartyProject` object, that is done with _one_ `cls(...)` call (inside the respective `.collect()` class methods), and I don't know how to inject/pass a `BaseProject` object into that construction.

If we were writing "regular" classes we would be in charge of writing the constructor ourselves, and we would probably end up with something like this:

```python
class BaseProject:
    @classmethod
    def parse_toml(toml_path: Path) -> TomlData:
        ...
        with toml_path.open("rb") as f:
            toml_data = tomllib.load(f)

    def __init__(self, toml_data: TomlData):
        self.name = toml_data["project"]["name"]
        self.description = toml_data["project"].get("description")
        self.experiments = ...

class SampleProject(BaseProject):
    def __init__(self, toml_data: TomlData):
        super().__init__(self, toml_data)
        self.path = ...
```

But since `dataclass` creates the constructor for us, we cannot pass around the `toml_data` in the same manner and extract the relevant information at each level of the class hierarchy.

In the end I settled on the structure you can see in `BaseExperiment.init_args_from_toml()`: We parse the relevant members from the TOML in the base class, but then we return them back to be subclass to then be passed to the `cls()` constructore. I see now that I failed to do the ~same clearer design for `BaseProject`. Will make another small refactoring here with extra docstrings to try to make things clearer.

Reply:
(crap, GitHub only showed me @vreuter's answer _after_ I had written the above in reply to @mknorps's comment)

Anyway, In b344f5b05a29ca1730f1fb256da069f704b08d02 I've tried to clean things up, and document the "protocol" here:
- The actual parsing of the TOML file is now in a `parse_toml()` helper function, as @vreuter suggests
- What remains in `BaseProject` is an `init_args_from_toml()` helper method, equivalent/corresponding to the one already in `BaseExperiment`. These helpers are indeed called from subclasses to delegate the extraction of the base class members from the parsed TOML data.

I added an underscore prefix as suggested by @vreuter in df87c66.